File: src/main/scala/com/bitdecay/tictactoe/scala/State.scala
package com.bitdecay.tictactoe.scala

import com.bitdecay.board.GameBoardState

case class State(q:Int=0,
                 w:Int=0,
                 e:Int=0,
                 a:Int=0,
                 s:Int=0,
                 d:Int=0,
                 z:Int=0,
                 x:Int=0,
                 c:Int=0,
                 winner:Int=0,
                 xTurn:Boolean=false,
                 lastPlayer:Int= -1,
                 finished:Boolean=false) extends GameBoardState {
  override lazy val serialize: String = s"$boardToString\n\n${State.print(winner)}, $xTurn, $finished, ${State.print(lastPlayer)}"
  private lazy val boardToString:String =
    s"""${State.print(q)}|${State.print(w)}|${State.print(e)}
       /-----
       /${State.print(a)}|${State.print(s)}|${State.print(d)}
       /-----
       /${State.print(z)}|${State.print(x)}|${State.print(c)}
     """.stripMargin('/')

  override def deserialize(str: String): GameBoardState = {
    val rows = str.split("\n")
    val values = rows.filter(_.contains("|")).flatMap(_.split("\\|"))
    val results = rows.last.split(", ")
    State(
      q=State.read(values(0)),
      w=State.read(values(0)),
      e=State.read(values(0)),
      a=State.read(values(0)),
      s=State.read(values(0)),
      d=State.read(values(0)),
      z=State.read(values(0)),
      x=State.read(values(0)),
      c=State.read(values(0)),
      winner=State.read(results(0)),
      xTurn=State.readBoolean(results(1)),
      finished=State.readBoolean(results(2)),
      lastPlayer=State.read(results(3))
    )
  }

  lazy val topRow = q + w + e
  lazy val midRow = a + s + d
  lazy val botRow = z + x + c
  lazy val leftCol = q + a + z
  lazy val midCol = w + s + x
  lazy val rightCol = e + d + c
  lazy val leftDiag = q + s + c
  lazy val rightDiag = e + s + z

  override lazy val toString = boardToString

  override lazy val clone:Object = copy()
}

object State {
  private def print(i:Int):String = i match {
    case -1 => "X"
    case 1 => "O"
    case _ => " "
  }

  private def read(s:String):Int = s.toLowerCase match {
    case "x" => -1
    case "o" => 1
    case _ => 0
  }

  private def readBoolean(s:String):Boolean = s.toLowerCase.equalsIgnoreCase("true")
}

File: src/main/java/com/bytebreakstudios/tictactoe/State.java
package com.bytebreakstudios.tictactoe;

import com.bytebreakstudios.board.GameBoardState;
import com.bytebreakstudios.board.utils.GameBoardException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class State implements GameBoardState {
    public int q = 0;
    public int w = 0;
    public int e = 0;
    public int a = 0;
    public int s = 0;
    public int d = 0;
    public int z = 0;
    public int x = 0;
    public int c = 0;
    public int winner = 0;
    public boolean xTurn = false;
    public int lastPlayer = -1;
    public boolean finished = false;

    private String print(int i){
        if (i == -1) return "X";
        else if (i == 1) return "O";
        else return " ";
    }

    private int read(String s){
        if (s.equalsIgnoreCase("X")) return -1;
        else if (s.equalsIgnoreCase("O")) return 1;
        else return 0;
    }

    private boolean readBoolean(String s){
        return s.equalsIgnoreCase("true");
    }

    @Override
    public String serialize() {
        StringBuilder sb = new StringBuilder();
        sb.append(boardToString()).append("\n\n");
        sb.append(print(winner)).append(", ").append(xTurn).append(", ").append(finished).append(", ").append(print(lastPlayer));
        return sb.toString();
    }

    private String boardToString(){
        StringBuilder sb = new StringBuilder();
        sb.append(print(q)).append("|").append(print(w)).append("|").append(print(e)).append("\n");
        sb.append("-----").append("\n");
        sb.append(print(a)).append("|").append(print(s)).append("|").append(print(d)).append("\n");
        sb.append("-----").append("\n");
        sb.append(print(z)).append("|").append(print(x)).append("|").append(print(c));
        return sb.toString();
    }

    @Override
    public GameBoardState deserialize(String str) {
        String[] rows = str.split("\n");
        List<String> vals = new ArrayList<>();
        for (String row : rows) if (row.contains("|")) Collections.addAll(vals, row.split("\\|"));
        q = read(vals.get(0));
        w = read(vals.get(1));
        e = read(vals.get(2));
        a = read(vals.get(3));
        s = read(vals.get(4));
        d = read(vals.get(5));
        z = read(vals.get(6));
        x = read(vals.get(7));
        c = read(vals.get(8));
        String[] results = rows[6].split(", ");
        this.winner = read(results[0]);
        this.xTurn = readBoolean(results[1]);
        this.finished = readBoolean(results[2]);
        this.lastPlayer = read(results[3]);
        return this;
    }

    public int topRow(){
        return q + w + e;
    }
    public int midRow(){
        return a + s + d;
    }
    public int botRow(){
        return z + x + c;
    }
    public int leftCol(){
        return q + a + z;
    }
    public int midCol(){
        return a + s + d;
    }
    public int rightCol(){ return e + d + c; }
    public int leftDiag(){
        return q + s + c;
    }
    public int rightDiag(){
        return e + s + z;
    }

    @Override
    public Object clone() {
        try {
            return super.clone();
        } catch (CloneNotSupportedException e) {
            throw new GameBoardException("This should never happen", e);
        }
    }

    @Override
    public String toString() {
        return boardToString();
    }
}
